Plugins/Data/NWScript/inc_array.nss

```diff
@@ -0,0 +1,485 @@
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    ExecuteStatement("DELETE FROM "+GetTableName(tag, obj)+" WHERE ind = " + IntToString(index), obj);
+    ExecuteStatement("UPDATE "+GetTableName(tag, obj)+" SET ind = ind - 1 WHERE ind > " + IntToString(index), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// return the index in the array containing "element"
+// if not found, return INVALID_INDEX
+int Array_Find_Str(string tag, string element, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    int tmp = -1;
+    string stmt = "SELECT MIN(ind) FROM "+GetTableName(tag, obj)+" WHERE value = @element";
+    sqlquery sqlQuery = SqlPrepareQueryObject(GetModule(), stmt);
+    SqlBindString(sqlQuery, "@element", element);
+    if ( SqlStep(sqlQuery) ) {
+        tmp = SqlGetInt(sqlQuery, 0);
+        WriteTimestampedLogEntry("Returned value = " + IntToString(tmp));
+        return tmp;
+    }
+    return INVALID_INDEX;
+}
+
+int Array_Find_Flt(string tag, float element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, FloatToString(element), obj);
+}
+
+int Array_Find_Int(string tag, int element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, IntToString(element), obj);
+}
+
+int Array_Find_Obj(string tag, object element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, ObjectToString(element), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Insert a new element into position 'index'.  If index is beyond the number of rows in the array,
+// this will quietly fail.  This could be changed if you wanted to support sparse
+// arrays.
+void Array_Insert_Str(string tag, int index, string element, object obj=OBJECT_INVALID)
+{
+    int rows = GetRowCount(tag, obj);
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    if (index < rows) {
+        ExecuteStatement("UPDATE "+GetTableName(tag, obj)+" SET ind = ind + 1 WHERE ind >= "+IntToString(index), obj);
+        ExecuteStatement("INSERT INTO "+GetTableName(tag, obj)+" VALUES ( " + IntToString(index) + ", '" + element + "')", obj);
+    }
+}
+
+void Array_Insert_Flt(string tag, int index, float element, object obj=OBJECT_INVALID)
+{
+    Array_Insert_Str(tag, index, FloatToString(element), obj);
+}
+
+void Array_Insert_Int(string tag, int index, int element, object obj=OBJECT_INVALID)
+{
+    Array_Insert_Str(tag, index, IntToString(element), obj);
+}
+
+void Array_Insert_Obj(string tag, int index, object element, object obj=OBJECT_INVALID)
+{
+    Array_Insert_Str(tag, index, ObjectToString(element), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Insert a new element at the end of the array.
+void Array_PushBack_Str(string tag, string element, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
```

Comment:
Why is each function doing this obj = GetModue(), if the leaf utility functions like GetRowCount/GetTableName will do it?